src/main/java11/org/kohsuke/github/extras/HttpClientGitHubConnector.java

```diff
@@ -22,16 +26,30 @@
  * {@link GitHubConnector} for {@link HttpClient}.
  *
  * @author Liam Newman
+ * @author Guillaume Smet
  */
 public class HttpClientGitHubConnector implements GitHubConnector {
 
+    private static final String AUTHORIZATION_HEADER = "Authorization";
+    private static final int MAX_REDIRECTS = 5;
+
     private final HttpClient client;
 
     /**
      * Instantiates a new HttpClientGitHubConnector with a default HttpClient.
      */
     public HttpClientGitHubConnector() {
-        this(HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NORMAL).build());
+        // We handle redirects manually as Java is copying all the headers when redirected
+        // even when redirecting to a different host which is problematic as we don't want
+        // to push the Authorization header when redirected to a different host.
+        // This problem was discovered when upload-artifact@v4 was released as the new
+        // service we are redirected to for downloading the artifacts doesn't support
+        // having the Authorization header set.
+        // The new implementation does not push the Authorization header when redirected
+        // to a different host, which is similar to what Okhttp is doing:
+        // https://github.com/square/okhttp/blob/f9dfd4e8cc070ca2875a67d8f7ad939d95e7e296/okhttp/src/main/kotlin/okhttp3/internal/http/RetryAndFollowUpInterceptor.kt#L313-L318
+        // See also https://github.com/arduino/report-size-deltas/pull/83 for more context
+        this(HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER).build());
```

Comment:
What about for private repos with uploaded artifacts?  Do they require the Authorization header after redirect there?

Reply:
Ah. That's an interesting question. I will test that. It would be a nightmare to handle so I hope not...

The URL is supposed to be extremely short lived so I think they suppose it is safe.